test/libsolidity/semanticTests/array/copying/copy_function_storage_array.sol

```diff
@@ -18,6 +18,6 @@ contract C {
 // compileViaYul: also
 // ----
 // test() -> 7
-// gas irOptimized: 126552
+// gas irOptimized: 124080
```

Comment:
This looks bad. Gas usage changed so this was generating code, just differently. Smells like a codegen error to me. I think we should investigate.

@chriseth What do you think? Results are still correct so could it really have been generating code that works incorrectly?

Reply:
The main fix in this PR is activating a special case that was previously always (or mostly) disabled because of an incorrect check. This reduces gas costs obviously.

Reply:
@chriseth But are we 100% sure that the code without this special case was correct? Was that just a bunch of extra conversion functions that would disappear when inlined?